src/replication/adapter/helmhub/chart_registry.go

```diff
@@ -0,0 +1,145 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package helmhub
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/goharbor/harbor/src/common/utils/log"
+	adp "github.com/goharbor/harbor/src/replication/adapter"
+	"github.com/goharbor/harbor/src/replication/model"
+	"github.com/pkg/errors"
+)
+
+func (a *adapter) FetchCharts(filters []*model.Filter) ([]*model.Resource, error) {
+	charts, err := a.client.fetchCharts()
+	if err != nil {
+		return nil, err
+	}
+
+	resources := []*model.Resource{}
+	repositories := []*adp.Repository{}
+	for _, chart := range charts.Data {
+		repository := &adp.Repository{
+			ResourceType: string(model.ResourceTypeChart),
+			Name:         chart.ID,
+		}
+		repositories = append(repositories, repository)
+	}
+
+	for _, filter := range filters {
+		if err = filter.DoFilter(&repositories); err != nil {
+			return nil, err
+		}
+	}
+
+	for _, repository := range repositories {
+		versionList, err := a.client.fetchChartDetail(repository.Name)
+		if err != nil {
+			log.Errorf("fetch chart detail: %v", err)
+			continue
```

Comment:
Should return the error directly, two reasons:
1. If we keep the process logic even if got an error already, the users don't know it in the portal, the replication is marked as succeed, but actually it is not.
2.  To keep consistence with other adapters

Reply:
ok